scripts/perf/scheduler_ops_metrics.py

```diff
@@ -0,0 +1,172 @@
+
+    1. Queuing delay - time taken from starting the executor to the task
+       instance to be added to the executor queue.
+    2. Start delay - time taken from starting the executor to the task instance
+       to start execution.
+    3. Land time - time taken from starting the executor to task instance
+       completion.
+    4. Duration - time taken for executing the task instance.
+
+    The DAGs implement bash operators that call the system wait command. This
+    is representative of typical operators run on Airflow - queries that are
+    run on remote systems and spend the majority of their time on I/O wait.
+
+    To Run:
+        $ python scripts/perf/scheduler_ops_metrics.py
+    """
+    __mapper_args__ = {
+        'polymorphic_identity': 'SchedulerMetricsJob'
+    }
+
+    def print_stats(self):
+        """
+        Print operational metrics for the scheduler test.
+        """
+        session = settings.Session()
+        TI = TaskInstance
+        tis = (
+            session
+            .query(TI)
+            .filter(TI.dag_id.in_(DAG_IDS))
+            .filter(TI.state.in_([State.SUCCESS]))
+            .all()
+        )
+        ti_perf = [(ti.dag_id, ti.task_id, ti.execution_date,
+                    (ti.queued_dttm - self.start_date).total_seconds(),
+                    (ti.start_date - self.start_date).total_seconds(),
+                    (ti.end_date - self.start_date).total_seconds(),
+                    ti.duration) for ti in tis]
+        ti_perf_df = pd.DataFrame(ti_perf, columns=['dag_id', 'task_id',
+                                                    'execution_date',
+                                                    'queue_delay',
+                                                    'start_delay', 'land_time',
+                                                    'duration'])
+
+        print('Performance Results')
+        print('###################')
+        for dag_id in DAG_IDS:
+            print('DAG {}'.format(dag_id))
+            print(ti_perf_df[ti_perf_df['dag_id'] == dag_id])
+        print('###################')
+
+        session.commit()
+
+    def heartbeat(self):
+        """
+        Override the scheduler heartbeat to determine when the test is complete
+        """
+        super(SchedulerMetricsJob, self).heartbeat()
+        session = settings.Session()
+        # Get all the relevant task instances
+        TI = TaskInstance
+        task_instances_to_examine = (
+            session
+            .query(TI)
+            .filter(TI.dag_id.in_(DAG_IDS))
+            .filter(TI.state.in_([State.SUCCESS]))
+            .all()
+        )
+        session.commit()
+
+        # Both DAGs have 4 tasks each and are executed for 5 days
+        if (len(task_instances_to_examine) == 40 or
+                (datetime.now()-self.start_date).total_seconds() >
+                MAX_RUNTIME_SECS):
+            self.logger.info("All tasks processed! Exiting.")
```

Comment:
Would be good to differentiate a timeout (> MAX_RUNTIME_SECS) and a regular completion.